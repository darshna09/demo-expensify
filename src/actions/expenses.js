import { v1 as uuidv1 } from 'uuid';

// Action Generator: ADD_EXPENSE
export const addExpense = ({
    description = '', note = '', amount = 0, createdAt = 0
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv1(),
        description,
        note,
        amount,
        createdAt
    }
});

// Action Generator: REMOVE_EXPENSE
export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// Action Generator: EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});