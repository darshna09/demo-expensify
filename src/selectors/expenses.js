// Start date, end date  - timestamp (milliseconds)
// Starts from Jnuary 1st 1970 from midnight (UNIX EPOCH)
// Positive milliseconds are after the above time while the ones in negative are before this time.

// Get visible expenses - filtered/sorted
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
