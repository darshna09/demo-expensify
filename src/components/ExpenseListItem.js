import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => (
    <div>
        <p>Description: {description}</p>
        <p>Amount: {amount}</p>
        <p>Created at: {createdAt}</p>
        <button onClick={() => {
            dispatch(removeExpense(id));
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem);