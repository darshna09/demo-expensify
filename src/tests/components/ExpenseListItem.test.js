import React from  'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('should render ExpenseListItem with expense', () => {
    const wrapper = shallow(<ExpenseListItem key={expenses[0].id} {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});
