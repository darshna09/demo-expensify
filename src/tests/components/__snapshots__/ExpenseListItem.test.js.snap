// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <h2>
    <Link
      replace={false}
      to="/edit/1"
    >
      Rent
    </Link>
  </h2>
  <p>
    Amount: 
    18750
  </p>
  <p>
    Created at: 
    1970-01-01
  </p>
</div>
`;
